src/preprocess.rkt

```diff
@@ -1,44 +1,122 @@
 #lang racket
 
-(require "syntax/types.rkt" "points.rkt" "float.rkt")
-
-(provide preprocess-pcontext *herbie-preprocess* apply-preprocess)
-
-;; Tracks list of preprocess structs Herbie decides to apply
-(define *herbie-preprocess* (make-parameter empty))
-
-;; index-value-pairs is a sorted list of (index, value)
-(define (list-set-multiple list index-value-pairs)
-  (let loop ([current list] [todo index-value-pairs] [index 0])
-    (cond
-      [(empty? current)
-       empty]
-      [(and (not (empty? todo)) (equal? (first (first todo)) index))
-       (cons (second (first todo)) (loop (rest current) (rest todo) (+ index 1)))]
-      [else
-       (cons (first current) (loop (rest current) todo (+ index 1)))])))
-
-(define (apply-to-group variables point group-variables group-function)
-  (define indicies
-    (map (lambda (var) (index-of variables var)) group-variables))
-  (define values
-    (group-function (map (curry list-ref point) indicies)))
-  (define sorted (sort (map list indicies values) (lambda (a b) (< (first a) (first b)))))
-  (list-set-multiple point sorted))
-
-(define (sort-group variables point preprocess repr)
-  (match-define (list 'sort vars ...) preprocess)
-  (apply-to-group variables point vars (lambda (group) (sort group (curryr </total repr)))))
-
-(define (apply-preprocess sampled-point preprocess-structs ctx)
+(require "core/egg-herbie.rkt" "core/simplify.rkt"
+         "syntax/syntax.rkt" "syntax/types.rkt" "alternative.rkt" "common.rkt"
+         "programs.rkt" "points.rkt" "timeline.rkt" "float.rkt")
+
+(provide find-preprocessing preprocess-pcontext remove-unnecessary-preprocessing)
+
+;; See https://pavpanchekha.com/blog/symmetric-expressions.html
+(define (find-preprocessing expression context rules)
```

Comment:
Ok, I think this is basically good and I'm willing to merge it. Longer term we should think about factoring out the individual preprocessing instructions, their identities, and so on from the glue code. But for now, with just two types, it's OK.

Reply:
Yeah I'm thinking along the same lines. We'll want some way to "define a preprocessing step," and then have driver code that uses them.